src/bm_sim/lpm_trie.cpp

```diff
@@ -0,0 +1,293 @@
+/* Copyright 2025 Contributors to the P4 Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <bm/bm_sim/_assert.h>
+#include <lpm_trie.h>
+
+namespace bm {
+
+struct Branch {
+  byte_t value;
+  std::unique_ptr<Node> next;
+};
+
+struct Prefix {
+  uint8_t prefix_length;
+  byte_t key;
+  value_t value;
+  Prefix(uint8_t prefix_length, byte_t key, value_t value)
+      : prefix_length(prefix_length), key(key), value(value) {}
+  bool operator<(const Prefix &other) const {
+    return (prefix_length == other.prefix_length)
+               ? (key < other.key)
+               : (prefix_length < other.prefix_length);
+  }
+};
+
+class Node {
+ public:
+  explicit Node(Node *parent = nullptr, byte_t child_id = 0)
+      : parent(parent), child_id(child_id) {}
+  Node(const Node &) = delete;
+
+  Node *get_next_node(byte_t byte) const;
+
+  void set_next_node(byte_t byte, std::unique_ptr<Node> next_node) {
+    add_branch(byte, std::move(next_node));
+  }
+
+  Node *get_parent() const { return parent; }
+
+  byte_t get_child_id() const { return child_id; }
+
+  // Prefix
+
+  bool get_prefix(uint8_t prefix_length, byte_t key, value_t *value);
+
+  void insert_prefix(uint8_t prefix_length, byte_t key, value_t value);
+
+  bool delete_prefix(uint8_t prefix_length, byte_t key);
+
+  // Branch
+  void add_branch(byte_t byte, std::unique_ptr<Node> nextNode);
+
+  bool delete_branch(byte_t byte);
+
+  bool is_empty() const { return prefixes.empty() && branches.empty(); }
+
+ private:
+  std::vector<Branch> branches;
+  std::vector<Prefix> prefixes;
+  Node *parent;
+  byte_t child_id;
+
+  std::vector<Prefix>::iterator search_prefix_with_key(uint8_t prefix_length,
+                                                       byte_t key);
+};
+
+Node *Node::get_next_node(byte_t byte) const {
+  auto it = std::lower_bound(
```

Comment:
@Hoooao This seems to fail in more recent versions of gcc:

https://github.com/p4lang/p4c/actions/runs/15490297907/job/43614080729?pr=5310#step:5:3194

Reply:
Unclear why the failure didn't occur here, maybe older CI?